tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,544 @@
+   */
+  coreIdToPos: function(id: number): number[] {
+    let p = this;
+    const chipId = Math.floor(id / 2);
+    const nodeId = id & 1;
+    const xDim = chipId % p.xDimension_;
+    const yDim = Math.floor(chipId / p.xDimension_);
+    const x =
+        p.getNodeXLoc(
+            Math.floor(xDim / p.hostXStride_), xDim % p.hostXStride_, nodeId) +
+        p.chipGridSize_ / p.nodesPerChip_ / 2;
+    const y = p.getChipYLoc(
+                  Math.floor(yDim / p.hostYStride_), yDim % p.hostYStride_) +
+        p.chipGridSize_ / 2;
+    return [x, y];
+  },
+  /**
+   * Returns the svg path given the src and dst core and node id.
+   * @return Path in svg format.
+   */
+  linkToPath: function(link): string {
+    let p = this;
+    const src = p.coreIdToPos(link.srcCoreId);
+    const dst = p.coreIdToPos(link.dstCoreId);
+    const path = 'M ' + src[0] + ' ' + src[1] + 'L ' + dst[0] + ' ' + dst[1];
+    return path;
+  },
+  /**
+   * Returns the text to visualize in the tool tips.
+   * @return String to render in tool tips.
+   */
+  getToolTipText_: function(data): string {
+    let parent = this;
+    let res = 'pos: (' + data.ydim + ',' + data.xdim + ')\n';
+    res += 'host: ' + data.host + '\n';
+    res += 'chip id: ' + data.cid + '\n';
+    res += 'node id: ' + data.nid + '\n';
+    res += 'replica id: ' + data.rid + '\n';
+    if (parent.selectedMetricIdx >= 0) {
+      res += parent.metrics[parent.selectedMetricIdx].label + ' spends ' +
+          data.value.toFixed(2) + 'us in total, ';
+      const pcnt = data.value / data.total * 100;
+      res += 'taking ' + pcnt.toFixed(2) + '% of a step.';
+    }
+    return res;
+  },
+  /**
+   * Draw the legend of the graph.
+   */
+  drawLegend: function(
+      svg: any, height: number, legendElementHeight: number,
+      colorScale: any, colors: number[]) {
+    const legendElementWidth = legendElementHeight * 2;
+    let legend = svg.selectAll('.legend').data(
+        [0].concat(colorScale.quantiles()), (d) => d);
+    let legend_g = legend.enter().append('g').attr('class', 'legend');
+    legend_g.append('rect')
+        .attr('x', (d, i) => legendElementWidth * i)
+        .attr('y', height)
+        .attr('width', legendElementWidth)
+        .attr('height', legendElementHeight)
+        .style('fill', (d, i) => colors[i]);
+    legend_g.append('text')
+        .text((d) => '\u2265 0.' + Math.round(d * 10))
+        .attr('x', (d, i) => legendElementWidth * i)
+        .attr('y', height + legendElementHeight * 2);
+    legend.exit().remove();
+  },
+  /**
+   * Updates the data to be loaded into the topology graph.
+   */
+  updateAllData: function(
+      data, runEnvironment, metrics, idx) {
+    if (!data || !runEnvironment || !runEnvironment.topology || !metrics ||
+        idx >= metrics.length || idx < 0) {
+      return;
+    }
+    const xdim = runEnvironment.topology.xDimension;
+    let result = [];
+
+    Object.keys(data.podStatsPerCore).forEach(function(val) {
```

Comment:
Few things:
1. `Object.keys(data.podStatsPerCore).forEach(function(val) {` -> ` Object.keys(data.podStatsPerCore).forEach(function(key) {`
2. You should consider using `map`.
```ts
this.topoData _ = Object.keys(data.podStatsPerCore).map((core) => {
    const podStats = data.podStatsPerCore[core];
     return {
        xdim: podStats.chipId % xdim,
        ydim: Math.floor(podStats.chipId / xdim),
        nid: podStats.nodeId,
        cid: podStats.chipId,
        rid: data.coreIdToReplicaIdMap[core], // replica id.
        host: podStats.hostName,
        value: podStats[metrics[idx].key],
        total: podStats.totalDurationUs
      });
});
```
3. since the data you create here can be flimsy, I would define a type for this object you are creating here.
4. `topoData_` should be a computed property.

Reply:
Done.